core/trino-main/src/main/java/io/trino/SystemSessionProperties.java

```diff
@@ -699,6 +705,21 @@ public SystemSessionProperties(
                         FAULT_TOLERANT_EXECUTION_TARGET_TASK_SPLIT_COUNT,
                         "Target number of splits for a single fault tolerant task",
                         queryManagerConfig.getFaultTolerantExecutionTargetTaskSplitCount(),
+                        false),
+                booleanProperty(
+                        ADAPTIVE_PARTIAL_AGGREGATION_ENABLED,
+                        "When enabled, partial aggregation can be skipped on an operator level if it's not beneficial",
+                        adaptivePartialAggregationConfig.isAdaptivePartialAggregationEnabled(),
+                        false),
+                longProperty(
+                        ADAPTIVE_PARTIAL_AGGREGATION_MIN_ROWS,
+                        "Minimum number of processed rows before partial aggregation can be skipped",
```

Comment:
> This will complicate the code a lot and only help with the unusual data distribution case where first you process a lot of unique groups and the switch to splits with a small number of groups.

How do you know it's unusual?
You can turn this argument around and say we hit a bad file and we turned off PA completely while the rest of data is rather flat.

In other places we turn adaptivness on/off: `DictionaryAwarePageProjectionWork#createDictionaryBlockProjection`. Generally, I think it's a preferred way because prefix of query might not be representative for the remainder of the query.

At very least we need to have a `TODO` and a plan for that